python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,742 @@
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import datetime
+import sys
+import uuid
+from decimal import ROUND_HALF_UP, Decimal
+from functools import singledispatch
+
+if sys.version_info >= (3, 8):
+    from functools import singledispatchmethod
+else:
+    from singledispatch import singledispatchmethod
+
+import pytz
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    Singleton,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+JAVA_MAX_INT = 2147483647
+JAVA_MIN_INT = -2147483648
+JAVA_MAX_FLOAT = 3.4028235e38
+JAVA_MIN_FLOAT = -3.4028235e38
+EPOCH = datetime.datetime.utcfromtimestamp(0)
+EPOCH_DAY = EPOCH.date()
+
+"""
+Iceberg literal is wrapper class used in expressions, which return unbound predicates
+It's being organized as below
+Literal
+|-- AboveMax
+|-- BelowMin
+|-- BaseLiteral
+    |-- StringLiteral
+    |-- FixedLiteral
+    |-- BinaryLiteral
+    |-- ComparableLiteral
+        |-- BooleanLiteral
+        |-- IntegerLiteral
+        |-- LongLiteral
+        |-- FloatLiteral
+        |-- DoubleLiteral
+        |-- DateLiteral
+        |-- TimeLiteral
+        |-- TimestampLiteral
+        |-- DecimalLiteral
+        |-- UUIDLiteral
+"""
+
+
+class Literal:
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def to_byte_buffer(self):
+        raise NotImplementedError()
+
+
+@singledispatch
+def of(value):
```

Comment:
Since this will be at the module level, is there a better name for this besides `of`? What do you think of `literal`? It feels right to do:
```py
from iceberg.expressions.literals import literal

l = literal(1.23)
print(l)  # FloatLiteral(value=1.23)
```

Reply:
sounds good, since capital `Literal` is used, I think we can use what you suggested instead